SPHINXsys/src/shared/particle_dynamics/relax_dynamics/relax_dynamics.cpp

```diff
@@ -238,6 +300,194 @@ namespace SPH
 			std::copy(GetParticles()->pos_n_.begin(), GetParticles()->pos_n_.end(), GetParticles()->pos_0_.begin());
 		}
 		//=================================================================================================//
+		ShellMidSurfaceBounding::
+			ShellMidSurfaceBounding(SPHBody &sph_body, NearShapeSurface &body_part, BaseBodyRelationInner &inner_relation, 
+				                    Real thickness, Real level_set_refinement_ratio) :
+			PartDynamicsByCell(sph_body, body_part), RelaxDataDelegateInner(inner_relation),
+			solid_particles_(dynamic_cast<SolidParticles*>(particles_)),
+			pos_n_(solid_particles_->pos_n_), n_0_(solid_particles_->n_0_),
+			constrained_distance_(0.5 * sph_body.sph_adaptation_->MinimumSpacing()),
+			particle_spacing_ref_(sph_body.sph_adaptation_->MinimumSpacing()),
+			thickness_(thickness), level_set_refinement_ratio_(level_set_refinement_ratio)
+		{
+			particles_->registerAVariable<Real>(color_, "Color");
+			particles_->registerAVariable<Vecd>(temporary_n_0_, "TemporatyNormal");
+			particles_->registerAVariable<Vecd>(previous_n_0_, "PreviousNormal");
+			level_set_shape_ = DynamicCast<LevelSetShape>(this, body_->body_shape_.getShapeByName(body_->getBodyName()));
+		}
+		//=================================================================================================//
+		void ShellMidSurfaceBounding::Update(size_t index_i, Real dt)
+		{
+			Vecd none_normalized_normal = level_set_shape_->findNoneNormalizedNormalDirection(pos_n_[index_i]);
+			Vecd normal = none_normalized_normal / (none_normalized_normal.norm() + TinyReal);
+			Real phi = level_set_shape_->findSignedDistance(pos_n_[index_i]);
+			Real relaxation_factor_ratio = particle_spacing_ref_ / level_set_refinement_ratio_ / (0.1 * thickness_);
+			Real factor = none_normalized_normal.norm() / relaxation_factor_ratio / thickness_;
+			pos_n_[index_i] -= factor * constrained_distance_ * normal;
+		}
+		//=================================================================================================//
+		void ShellMidSurfaceBounding::getNormalDirection()
+		{
+			parallel_for(blocked_range<size_t>(0, particles_->total_real_particles_),
+				[&](const blocked_range<size_t>& r) {
+				for (size_t i = r.begin(); i != r.end(); ++i)
+				{
+					n_0_[i] = level_set_shape_->findNormalDirection(pos_n_[i] + 0.2 * thickness_ * n_0_[i]);
+					n_0_[i] /= n_0_[i].norm() + TinyReal;
+				}
+			}, ap);
+		}
+		//=================================================================================================//
+		void ShellMidSurfaceBounding::setColorFunction()
+		{
+			parallel_for(blocked_range<size_t>(0, particles_->total_real_particles_),
+				[&](const blocked_range<size_t>& r) {
+				for (size_t i = r.begin(); i != r.end(); ++i)
+				{
+					color_[i] = 0;
+				}
+			}, ap);
+			color_[0] = 1;
+		}
+		//=================================================================================================//
+		void ShellMidSurfaceBounding::correctNormalDirection()
```

Comment:
@Xiangyu-Hu can we use parallel loops here too? @mrezavandVirtonomy

Reply:
yes. Dong can do it.